File: jarvis-desktop/src/card-words.js
/**
 * The words on an approval card, and what Jarvis says aloud about a card -
 * the same in the Jarvis bar, the widget, the HUD page and the phone.
 *
 * The creativity audit (2026-09-25, docs/creativity-2026-09-25/experience.md
 * finding 1) found the card speaking three dialects: the bar showed the
 * code name (`switch_model`), the widget "APPROVAL REQUIRED", the phone
 * "Jarvis wants to learning enable", and Approve and Deny swapped places.
 * Now the PC writes every title (backend/jarvis_card_words.py, served as
 * each row's `notice.title`), and every screen shows:
 *
 *   Needs your OK                 <- CARD_KICKER
 *   Jarvis wants to search the web  <- the title (cardTitle)
 *   ...the rest of the card...
 *   [Deny]  [Approve]             <- Deny left, Approve right (CARD_BUTTONS)
 *
 * tests/card-words.mjs holds every word here to
 * tests/fixtures/card-words-cases.json, which tools/gen_card_words_cases.py
 * makes from the backend - the phone's CardWordsContractTest reads the same
 * file, so the two apps cannot word a card differently.
 *
 * No page, no Tauri: node can import it.
 *
 * @module card-words
 */

/** The small label above every card's title. */
export const CARD_KICKER = "Needs your OK";

/** The two buttons, left to right, on every screen. Why this order:
 *  docs/ARCHITECTURE.md §3, "One card on every screen". */
export const CARD_BUTTONS = Object.freeze(["Deny", "Approve"]);

/** The title of a row that names no action at all. */
export const CARD_NO_ACTION = "Jarvis is asking for your approval";

/**
 * What Jarvis says during a SPOKEN question that waits on a card, and
 * afterwards - keyed by the chat stream's `: jarvis-status` word
 * (backend/jarvis_agent.py: "approval", then the gate's outcome). Fixed
 * sentences: nothing from the card, so safe in any room. There is no
 * approving by voice; these only say where the card is.
 */
export const CARD_VOICE = Object.freeze({
  waiting: "I need your OK for that. There's a card on your screen.",
  approved: "Approved. Carrying on.",
  denied: "OK, I won't do that.",
  timed_out: "That card timed out, so nothing was done.",
});

const CARD_LINES = new Set(Object.values(CARD_VOICE));

/** True for one of the fixed card lines above (never an answer's words). */
export function isCardLine(text) {
  return CARD_LINES.has(text);
}

/**
 * A title for a row with no `notice` (a backend older than
 * approval-notice.patch): the action's name, quoted as a name - the PC's own
 * fallback for an action it has no phrase for (jarvis_card_words.title_for),
 * letter for letter. Never from `prompt` or `detail`: a title also ends up
 * in a Windows notification.
 */
export function fallbackTitle(action) {
  const raw = typeof action === "string" ? action : "";
  const words = raw
    .replace(/_/g, " ")
    .replace(/[^A-Za-z0-9 ]+/g, " ")
    .split(" ")
    .filter(Boolean)
    .join(" ")
    .slice(0, 60)
    .trim();
  return words ? `Jarvis wants your OK for "${words}"` : CARD_NO_ACTION;
}

/** The title every surface shows: the notice's, else the fallback. */
export function cardTitle(approval) {
  if (!approval || typeof approval !== "object") return CARD_NO_ACTION;
  if (typeof approval.title === "string" && approval.title.trim()) return approval.title.trim();
  const notice = approval.notice;
  const title = notice && typeof notice.title === "string" ? notice.title.trim() : "";
  return title || fallbackTitle(approval.action);
}

/**
 * Which card line to say, as the chat stream's status words arrive during
 * ONE spoken question. "approval" says the waiting line once per card; an
 * outcome word says what happened - only after the waiting line was said,
 * so an outcome never arrives out of nowhere. Everything else says nothing.
 * The phone's `CardVoice` does exactly this (the `voice_script` cases).
 */
export function createCardVoice() {
  let waiting = false;
  return {
    /** A new question: nothing said yet. */
    reset() {
      waiting = false;
    },
    /** @returns {string|null} the line to say, or null. */
    onStatus(word) {
      if (word === "approval") {
        if (waiting) return null;
        waiting = true;
        return CARD_VOICE.waiting;
      }
      if (word === "approved" || word === "denied" || word === "timed_out") {
        if (!waiting) return null;
        waiting = false;
        return CARD_VOICE[word];
      }
      return null;
    },
  };
}

File: jarvis-client/app/src/main/java/com/jarvis/client/net/CardWords.kt
package com.jarvis.client.net

/**
 * The words on an approval card, and what Jarvis says aloud about a card -
 * the same on this phone as in the desktop's Jarvis bar, widget and HUD.
 *
 * The creativity audit (2026-09-25, docs/creativity-2026-09-25/experience.md
 * finding 1) found the card speaking three dialects - this phone said
 * "Jarvis wants to learning enable", the desktop showed `switch_model`, and
 * Approve and Deny swapped places between screens. Now the PC writes every
 * title (backend/jarvis_card_words.py, served as each row's `notice.title`)
 * and every screen shows the same layout:
 *
 *     Needs your OK                     <- [KICKER]
 *     Jarvis wants to search the web    <- the title
 *     ...the rest of the card...
 *     [Deny]  [Approve]                 <- Deny left, Approve right ([BUTTONS])
 *
 * Deny on the left and Approve on the right matches this card's own swipe
 * (right approves, left denies) and Android's own dialogs, which put the
 * confirming button on the right. docs/ARCHITECTURE.md §3 has the reasons.
 *
 * CardWordsContractTest holds every word here to
 * `contract/card-words-cases.json`, which tools/gen_card_words_cases.py makes
 * from the backend - the desktop's tests/card-words.mjs reads the same file.
 */
object CardWords {
    /** The small label above every card's title. */
    const val KICKER = "Needs your OK"

    /** The two buttons, left to right, on every screen. */
    val BUTTONS = listOf("Deny", "Approve")

    /** The title of a row that names no action at all. */
    const val NO_ACTION = "Jarvis is asking for your approval"

    /** "Open the card" - the line on every screen but Home while a card waits. */
    const val OPEN_CARD = "Open the card"

    /**
     * What Jarvis says during a SPOKEN question that waits on a card, and
     * afterwards - keyed by the chat stream's `: jarvis-status` word
     * (backend/jarvis_agent.py: "approval", then the gate's outcome). Fixed
     * sentences: nothing from the card. There is no approving by voice: the
     * voice check cannot tell a recording from the owner, so these only say
     * where the card is.
     */
    val VOICE: Map<String, String> = mapOf(
        "waiting" to "I need your OK for that. There's a card on your screen.",
        "approved" to "Approved. Carrying on.",
        "denied" to "OK, I won't do that.",
        "timed_out" to "That card timed out, so nothing was done.",
    )

    /** What the "Open the card" line shows: which card, and how many more. */
    data class Waiting(val id: String, val title: String, val more: String)

    /**
     * The "Open the card" line for [items], or null when nothing waits. The
     * NEWEST card is named - the last one read, most likely the one the
     * owner's last tap raised - and "and N more waiting" says the rest are
     * there too. The desktop's `cardLinkView` (card-link.js) says the same.
     */
    fun waiting(items: List<PendingItem>): Waiting? {
        val newest = items.lastOrNull() ?: return null
        val more = items.size - 1
        return Waiting(newest.id, newest.title, if (more > 0) "and $more more waiting" else "")
    }

    private val NOT_A_WORD = Regex("[^A-Za-z0-9 ]+")

    /**
     * A title for a row with no `notice` (a backend older than
     * approval-notice.patch): the action's name, quoted as a name - the PC's
     * own fallback for an action it has no phrase for, letter for letter.
     * Never from `prompt` or `detail`: a title also ends up on a lock screen.
     */
    fun fallbackTitle(action: String?): String {
        val words = (action ?: "")
            .replace('_', ' ')
            .replace(NOT_A_WORD, " ")
            .split(' ')
            .filter { it.isNotEmpty() }
            .joinToString(" ")
            .take(60)
            .trim()
        return if (words.isEmpty()) NO_ACTION else "Jarvis wants your OK for \"$words\""
    }
}
